### DISCUSSION TITLE
CborReader requires a ReadOnlyMemory<byte>

### DISCUSSION THREAD

**Discussion by chrisspre** [dotnet/runtime#73022] [General] [answered]
CborReader has only one [constructor ](https://github.com/dotnet/runtime/blob/d7c8bc10ac00fe330daa633f8a2b16ba208e6e56/src/libraries/System.Formats.Cbor/src/System/Formats/Cbor/Reader/CborReader.cs#L56) and that requires a ReadOnlyMemory<byte> 

Would it not be possible to pass in a ReadOnlySpan<byte> and therefore saving a potential allocation for the client of the reader?

**Comment by bartonjs** [comment, marked as answer]
CborReader can't save a `ReadOnlySpan<byte>` as a field because it is not a `ref struct`.

When we built it, we felt that a class is much more usable, and that there probably weren't many scenarios where the caller would have a span versus an array (or Memory)

**Reply by chrisspre** [reply]
Thanks Jeremy. I did not realize immediately that the Reader keeps a reference to the data. But it makes sense that it does.

**Reply by bartonjs** [reply]
Yeah, in a constructor `ReadOnlySpan<byte>` means "I make a copy" (unless it's a ref struct) and `ReadOnlyMemory<byte>` means "I hold a reference".  It's nicer over arrays where it was sort of arbitrary of reference vs copy.

**Reply by chrisspre** [reply]
Understood.

It might be nice in the long run (for some advanced scenarios) to have a stateless CBOR "decoder" separate from the reader that I can point at an arbitrary span  and it gives me back the "token"  
e.g.  `static int ReadInt32(ReadOnlySpan<byte> span)` 
For cases where I own the byte array/span, know the position of the CBOR element and don't require the reader to scan over the byte array to get to that position and track the state (depth, nesting, etc. )
I think the code for this is already in the reader but all internal (PeekSignerInteger, AdvanceDataItemCounters, ...)

**Reply by c-ohle** [reply]
I avoid such problems by using ref Span's:
`static int ReadInt32(ref ReadOnlySpan<byte> span)`
